packages/monaco/src/browser/monaco-quick-open-service.ts

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from 'inversify';
+import { QuickOpenService, QuickOpenModel, QuickOpenOptions, QuickOpenItem, QuickOpenGroupItem } from "@theia/core/lib/browser";
+
+export interface InternalMonacoQuickOpenOptions extends monaco.quickOpen.IQuickOpenControllerOpts {
+    readonly prefix?: string;
+    onClose?(canceled: boolean): void;
+}
+
+@injectable()
+export class MonacoQuickOpenService extends QuickOpenService {
+
+    protected _widget: monaco.quickOpen.QuickOpenWidget | undefined;
+    protected options: InternalMonacoQuickOpenOptions | undefined;
+
+    open(model: QuickOpenModel, options?: QuickOpenOptions): void {
+        this.internalOpen(new MonacoQuickOpenOptions(model, options));
+    }
+
+    internalOpen(options: InternalMonacoQuickOpenOptions): void {
+        this.options = options;
+        this.widget.show(this.options.prefix || '');
+    }
+
+    protected get widget(): monaco.quickOpen.QuickOpenWidget {
+        if (this._widget) {
+            return this._widget;
+        }
+        const overlayWidgets = document.createElement('div');
+        overlayWidgets.classList.add('quick-open-overlay');
+        document.body.appendChild(overlayWidgets);
+
+        const container = document.createElement('quick-open-container');
```

Comment:
This service uses the DOM ?
So it won't be testable in node right ?

Should that be in another widget ?

Reply:
this code is responsible for managing `QuickOpenWidget` from vscode which heavily depends on DOM, `QuickOpenWidget` is already should be tested in vscode

To add a new quick service, one should implement `QuickOpenModel` from the core and it is possible to test it with pure Node.js tests.

Reply:
Do we need a library like https://github.com/rstacruz/mocha-jsdom to tests stuff like document.createElement in a node environment? Maybe it works even without? I haven't really tried but I was just wondering as I think this could simplify testing frontend stuff if we just do it in node.

Reply:
I meant that the maybe the quick-open-container' should not be in the service directly?